README.md
# grunt-aws-s3-gzip

> Interact with AWS S3 using AWS SDK

This is a fork of [grunt-aws-s3](https://github.com/MathieuLoutre/grunt-aws-s3) with added GZip support and other minor improvements.

## Warning 

Versions 0.4.0 to 0.5.0 have a bug where `options.params` is ignored.  
Version 0.8.0 doesn't actually support Node 0.8.x and 0.9.x.

## Getting Started
This plugin requires Grunt `~0.4.0`

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
  npm install grunt-aws-s3-gzip --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
  grunt.loadNpmTasks('grunt-aws-s3-gzip');
```

## The "aws_s3" task

### Options

#### options.accessKeyId (required)
Type: `String`

The AWS accessKeyId. You can load it via JSON as shown in the example or use the `AWS_ACCESS_KEY_ID` environment variable.

#### options.secretAccessKey (required)
Type: `String`

The AWS secretAccessKey. You can load it via JSON as shown in the example or use the `AWS_SECRET_ACCESS_KEY` environment variable.

#### options.bucket (required)
Type: `String`

The AWS bucket name you want to upload to.

#### options.region
Type: `String`  
Default: `US Standard`

The AWS [region](http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region).

If not specified, it uploads to the default 'US Standard'

#### options.endpoint
Type: `String`  

The AWS [endpoint](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Endpoint.html).

If not specified, it uploads to the default endpoint provided by the AWS Javascript SDK.

#### options.maxRetries
Type: `Integer`

The maximum amount of retries to attempt with a request.

#### options.sslEnabled
Type: `Boolean`  

Whether to enable SSL for requests or not.

#### options.httpOptions
Type: `Object`  

A set of options to pass to the low-level HTTP request. The list of option can be found in the [documentation](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#constructor-property)

#### options.access
Type: `String`  
Default:`public-read`

The ACL you want to apply to ALL the files that will be uploaded. The ACL values can be found in the [documentation](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property).

#### options.uploadConcurrency
Type: `Integer`  
Default: `1`

Number of uploads in parallel. By default, there's no concurrency. Must be > 0.
Note: This used to be called `concurrency` but the option has been deprecated, however it is still backwards compatible until 1.0.0.

#### options.downloadConcurrency
Type: `Integer`  
Default: `1`

Number of download in parallel. By default, there's no concurrency. Must be > 0.

#### options.params
Type: `Object`

A hash of the params you want to apply to the files. Useful to set the `ContentEncoding` to `gzip` for instance, or set the `CacheControl` value. The list of parameters can be found in the [documentation](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property). `params` will apply to *all* the files in the target. However, the `params` option in the file list has priority over it.

#### options.mime
Type: `Object`  

The MIME type of every file is determined by a MIME lookup using [node-mime](https://github.com/broofa/node-mime). If you want to override it, you can use this option object.
The keys are the local file paths and the values are the MIME types.

```js
  {
    'path/to/file': 'application/json',
    'path to/other/file': 'application/gzip'
  }
```

You need to specify the full path of the file, including the `cwd` part.  
The `mime` hash has absolute priority over what has been set in `options.params` and the `params` option of the file list.

#### options.stream
Type: `Boolean`  
Default: `false`

Allows to use streams instead of buffers to upload and download.
The option can either be turned on for the whole subtask or for a specified file object like so:

```js
  {'action': 'upload', expand: true, cwd: 'dist/js', src: ['**'], stream: true}
```

#### options.debug
Type: `Boolean`  
Default: `false`

This will do a "dry run". It will not upload anything to S3 but you will get the full report just as you would in normal mode. Useful to check what will be changed on the server before actually doing it. Unless one of your actions depends on another (like download following a delete), the report should be accurate.  
`listObjects` requests will still be made to list the content of the bucket.

#### options.differential
Type: `Boolean`  
Default: `false`

`listObjects` requests will be made to list the content of the bucket, then they will be checked against their local file equivalent (if it exists) using MD5 (and sometimes date) comparisons.
This means different things for different actions:
- `upload`: will only upload the files which either don't exist on the bucket or have a different MD5 hash
- `download`: will only download the files which either don't exist locally or have a different MD5 hash and are newer.
- `delete`: will only delete the files which don't exist locally

The option can either be specified for the whole subtask or for a specified file object like so:

```js
  {'action': 'upload', expand: true, cwd: 'dist/js', src: ['**'], differential: true}
```

In order to be able to compare to the local file names, it is necessary for `dest` to be a finished path (e.g `directory/` instead of just `dir`) as the comparison is done between the file names found in `cwd` and the files found on the server `dest`. If you want to compare the files in the directory `scripts/` in your bucket and the files in the corresponding local directory `dist/scripts/` you need to have something like:

```js
  {cwd: 'dist/scripts/', dest: 'scripts/', 'action': 'download', differential: true}
```

#### options.displayChangesOnly
Type: `Boolean`

Default: `false`

If enabled, only lists files that have changed when performing a differential upload.

#### options.gzip
Type: `Boolean`

Default: `true`

Compress files on the fly using gzip and add the `ContentEncoding: gzip` header.

#### options.excludedFromGzip
Type: `Array<String>` or `String`

Default: None

If `options.gzip` is specified, files matching this pattern will not be compressed.

```js
{ gzip: true, excludedFromGzip: ['*.png', '*.jpg', '*.jpeg'] }
```

### Actions

This Grunt task supports three modes of interaction with S3, `upload`, `download` and `delete`. Every action that you specify is executed serially, one after the other. If multiple `upload` actions are one after the other, they will be grouped together.

You choose the action by specifying the key `action` in the file hash like so:

```js
  {'action': 'upload', expand: true, cwd: 'dist/js', src: ['**'], dest: 'app/js/'}
```

By default, the action is `upload`.

#### `upload`

The `upload` action uses the [newest Grunt file format](http://gruntjs.com/configuring-tasks#files), allowing to take advantage of the `expand` and `filter` options.  
It is the default action, so you can omit `action: 'upload'` if you want a cleaner look.

```js
  files: [
    {expand: true, cwd: 'dist/staging/scripts', src: ['**'], dest: 'app/scripts/'},
    {expand: true, cwd: 'dist/staging/styles', src: ['**'], dest: 'app/styles/', action: 'upload'}
  ]
```

You can also include a `params` hash which will override the options.params one. For example:

```js
  
  params: {
    ContentType: 'application/json'
    CacheControl: '3000'
  }

  // ...

  files: [
    {expand: true, cwd: 'dist/staging/scripts', src: ['**'], dest: 'app/scripts/', params: {CacheControl: '2000'}},
    {expand: true, cwd: 'dist/staging/styles', src: ['**'], dest: 'app/styles/'}
  ]
```

This will yield for the params which will eventually be applied:

```js
  {
    ContentType: 'application/json',
    CacheControl: '2000'
  }

  // AND

  {
    ContentType: 'application/json',
    CacheControl: '3000'
  }
```

The `options.mime` hash, however, has priority over the ContentType. So if the hash looked like this:

```js
  {
    'dist/staging/styles/LICENCE': 'text/plain'
  }
```

The `ContentType` eventually applied to `dist/staging/styles/LICENCE` would be `text/plain` even though we had a `ContentType` specified in `options.params` or in `params` of the file.

When the `differential` option is enabled, it will only upload the files which either don't exist on the bucket or have a different MD5 hash.

#### `download`

The `download` action requires a `cwd`, a `dest` and *no* `src` like so:

```js
  {cwd: 'download/', dest: 'app/', action: 'download'}
```

The `dest` is used as the Prefix in the [listObjects command](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#listObjects-property) to find the files _on the server_ (which means it can be a path or a partial path). 
The `cwd` is used as the root folder to write the downloaded files. The inner folder structure will be reproduced inside that folder.

If you specify '/' for `dest`, the whole bucket will be downloaded. It handles automatically buckets with more than a 1000 objects.  
If you specify 'app', all paths starting with 'app' will be targeted (e.g. 'app.js', 'app/myapp.js', 'app/index.html, 'app backup/donotdelete.js') but it will leave alone the others (e.g. 'my app/app.js', 'backup app/donotdelete.js').

When the `differential` options is enabled, it will only download the files which either don't exist locally or have a different MD5 hash and are newer. 

Note: if `dest` is a file, it will be downloaded to `cwd` + `file name`. If `dest` is a directory ending with `/`, its content will be downloaded to `cwd` + `file names or directories found in dest`. If `dest` is neither a file nor a directory, the files found using it as a prefix will be downloaded to `cwd` + `paths found using dest as the prefix`.

The `download` action can also take an `exclude` option like so:

```js
  {cwd: 'download/', dest: 'app/', action: 'download', exclude "**/.*"}
```

The value is a globbing pattern that can be consumed by `grunt.file.isMatch`. You can find more information on [globbing patterns on Grunt's doc](http://gruntjs.com/api/grunt.file#globbing-patterns). In this example, it will exclude all files starting with a `.` (they won't be downloaded).
If you want to reverse the `exclude` (that is, only what will match the pattern will be downloaded), you can use the `flipExclude` option like so:

```js
  {cwd: 'download/', dest: 'app/', action: 'download', exclude "**/.*", flipExclude: true}
```


CHANGELOG.md
# CHANGELOG

### v0.8.6
- New option to set endpoint for such use as integrating with minio server.

### v0.8.5 
- New option to display changes only

### v0.8.4
- Fix bug in setImmediate support detection (was using the shim even on Node 0.10.x)

### v0.8.2-3
- Unpublished because the attempt to better detect setImmediate was faulty

### v0.8.1
- Shim setImmediate to support Node 0.8.x

### v0.8.0
- If a directory is found during download, it will be skipped (and won't create empty dirs). This happened only a an empty directory has been created manually on S3 (by @nicolindemann)
- Use a glob pattern to exclude files when downloading a folder (with @nicolindemann)
- Chaneg dot colour wether the object has been downloaded/uploaded
- Exclude option for delete
- Refactor
- Stream option (with @craigloftus)
- Bug fixes on options priority (if you had differential set to true for the whole task/subtask but to false for a file object, true would take priority)

### v0.7.2

- Follow Grunt 0.4.2 guidelines and include external `lodash` and `async`
- Add more options (httpOptions, maxRetries, sslEnabled)
- Fix a bug when downloading and extra keys where sent to getObject
- Fix download of a single item. Key paths are now relative to given dest if dest is a directory or the file itself

### v0.7.1

- If a marker is not provided when listing objects but the list is flagged as truncated, use last element as marker (by @derekr)

### v0.7.0

- Grouping of uploads together (treated like `delete` and `download` action in the code)
- Improved log
- Debug option to do dry runs
- Differential upload based on MD5 checks (using S3's ETags)
- Differential download based on wether it exists locally or not, MD5 checks and date
- Differential delete based on wether it still exists locally or not
- Tests using `mock-aws-s3` to replace the AWS package during testing
- Code restructure/formatting
- Update docs

### v0.6.0

- Add 'download' option.
- Fix `options.params` not being applied
- Add a `params` option field to the file hash which overrides `options.params`
- The `mime` hash has priority over the `params` option field of the file
- Multiple code style/lint fixes
- Remove uploading of empty directories
- Nicer log
- Add changelog!
- Better documentation

### v0.5.0

- Add option to override automatic MIME type detection

### v0.4.1

- Fix delete task executing separately from upload

### v0.4.0

- Add 'delete' option.
- _Breaks the use of `options.params`_

### v0.3.1

- Region is now optional, defaults to US Standard

### v0.3.0

- Option for upload concurrency.

### v0.2.0

- Can set additional params and bug fix

### v0.1.1

- Fix bug when using env variable.

### v0.1.0

- First release. Simple upload to S3.


tasks/aws_s3.js
/*
 * grunt-aws-s3
 * https://github.com/MathieuLoutre/grunt-aws-s3
 *
 * Copyright (c) 2013 Mathieu Triay
 * Licensed under the MIT license.
 */

'use strict';

var path = require('path');
var fs = require('fs');
var crypto = require('crypto');
var AWS = require('aws-sdk');
var mime = require('mime');
var _ = require('lodash');
var async = require('async');
var zlib = require('zlib');

require('setimmediate'); // For compatibility with Node 0.8.x

module.exports = function (grunt) {

	grunt.registerMultiTask('aws_s3', 'Interact with AWS S3 using the AWS SDK', function () {

		var done = this.async();
		var options = this.options({
			access: 'public-read',
			accessKeyId: process.env.AWS_ACCESS_KEY_ID,
			secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
			uploadConcurrency: 1,
			downloadConcurrency: 1,
			mime: {},
			params: {},
			debug: false,
			mock: false,
			differential: false,
			stream: false,
			displayChangesOnly: false,
			gzip: true,
			excludedFromGzip: []
		});
		
		// To deprecate
		if (options.concurrency !== undefined) {
			grunt.log.writeln('The concurrency option is deprecated, use uploadConcurrency instead\n'.yellow);
			options.uploadConcurrency = options.concurrency;
		}
		if (options.s3)

			var filePairOptions = {
				differential: options.differential,
				stream: options.stream,
				gzip: options.gzip,
				excludedFromGzip: options.excludedFromGzip,
				flipExclude: false,
				exclude: false
			};

		// Replace the AWS SDK by the mock package if we're testing
		if (options.mock) {
			AWS = require('mock-aws-s3');
		}

		// List of acceptable params for an upload
		var put_params = ['CacheControl', 'ContentDisposition', 'ContentEncoding',
			'ContentLanguage', 'ContentLength', 'ContentMD5', 'Expires', 'GrantFullControl',
			'GrantRead', 'GrantReadACP', 'GrantWriteACP', 'Metadata', 'ServerSideEncryption',
			'StorageClass', 'WebsiteRedirectLocation', 'ContentType'
		];

		// Checks that all params are in put_params
		var isValidParams = function (params) {

			return _.every(_.keys(params), function (key) {
				return _.contains(put_params, key);
			});
		};

		var getObjectURL = function (file) {

			file = file || '';
			return s3.endpoint.href + options.bucket + '/' + file;
		};

		// Get the key URL relative to a path string 
		var getRelativeKeyPath = function (key, dest) {

			var path;

			if (_.last(dest) === '/') {
				// if the path string is a directory, remove it from the key
				path = key.replace(dest, '');
			} else if (key.replace(dest, '') === '') {
				path = _.last(key.split('/'));
			} else {
				path = key;
			}

			return path;
		};

		var hashFile = function (options, callback) {
			var hash = crypto.createHash('md5');

			function getHashString() {
				// S3's ETag has quotes around it...
				return '"' + hash.digest('hex') + '"';
			}

			if (options.stream) {
				var local_stream = fs.ReadStream(options.file_path);
				if (options.gzip) {
					local_stream = local_stream.pipe(zlib.createGzip());
				}

				local_stream.on('end', function () {
					callback(null, getHashString());
				});

				local_stream.on('error', function (err) {
					callback(err);
				});

				local_stream.on('data', function (data) {
					hash.update(data);
				});
			} else {
				var local_buffer = grunt.file.read(options.file_path, {
					encoding: null
				});

				var getHashStringForBuffer = function getHashStringForBuffer(buffer) {
					hash.update(buffer);
					return getHashString();
				};

				if (options.gzip) {
					zlib.gzip(local_buffer, function (err, compressed) {
						callback(err, err ? null : getHashStringForBuffer(compressed));
					});
				} else {
					callback(null, getHashStringForBuffer(local_buffer));
				}
			}
		};

		// Checks that local file is 'date_compare' than server file
		var checkFileDate = function (options, callback) {

			fs.stat(options.file_path, function (err, stats) {

				if (err) {
					callback(err);
				} else {
					var local_date = new Date(stats.mtime).getTime();
					var server_date = new Date(options.server_date).getTime();

					if (options.compare_date === 'newer') {
						callback(null, local_date > server_date);
					} else {
						callback(null, local_date < server_date);
					}
				}
			});
		};

		var isFileDifferent = function (options, callback) {

			hashFile(options, function (err, md5_hash) {

				if (err) {
					callback(err);
				} else {
					if (md5_hash === options.server_hash) {
						callback(null, false);
					} else {
						if (options.server_date) {
							options.compare_date = options.compare_date || 'older';
							checkFileDate(options, callback);
						} else {
							callback(null, true);
						}
					}
				}
			});
		};

		if (!options.accessKeyId && !options.mock) {
			grunt.warn('Missing accessKeyId in options');
		}

		if (!options.secretAccessKey && !options.mock) {
			grunt.warn('Missing secretAccessKey in options');
		}

		if (!options.bucket) {
			grunt.warn('Missing bucket in options');
		}

		
		var s3_options = {
			bucket: options.bucket,
			accessKeyId: options.accessKeyId,
			secretAccessKey: options.secretAccessKey
		};
		
		if (options.endpoint) {
			var ep = new AWS.Endpoint(options.endpoint);
			s3_options.endpoint = ep;
		} 

		if (!options.region) {
			grunt.log.writeln('No region defined. S3 will default to US Standard\n'.yellow);
		} else {
			s3_options.region = options.region;
		}

		if (options.params) {
			if (!isValidParams(options.params)) {
				grunt.warn('"params" can only be ' + put_params.join(', '));
			}
		}

		// Allow additional (not required) options
		_.extend(s3_options, _.pick(options, ['maxRetries', 'sslEnabled', 'httpOptions']));

		var s3 = new AWS.S3(s3_options);

		var dest;
		var is_expanded;
		var objects = [];
		var uploads = [];

		// Because Grunt expands the files array automatically, 
		// we need to group the uploads together to make the difference between actions.
		var pushUploads = function () {

			if (uploads.length > 0) {
				objects.push({
					action: 'upload',
					files: uploads
				});
				uploads = [];
			}
		};

		this.files.forEach(function (filePair) {

			is_expanded = filePair.orig.expand || false;

			if (filePair.action === 'delete') {

				_.defaults(filePair, filePairOptions);

				if (!filePair.dest && filePair.dest !== '') {
					grunt.fatal('No "dest" specified for deletion. No need to specify a "src"');
				} else if (filePair.differential && !filePair.cwd) {
					grunt.fatal('Differential delete needs a "cwd"');
				}

				pushUploads();

				filePair.dest = (filePair.dest === '/') ? '' : filePair.dest;

				objects.push(filePair);
			} else if (filePair.action === 'download') {

				if (is_expanded) {
					grunt.fatal('You cannot expand the "src" for downloads');
				} else if (!filePair.dest) {
					grunt.fatal('No "dest" specified for downloads');
				} else if (!filePair.cwd || filePair.src) {
					grunt.fatal('Specify a "cwd" but not a "src" for downloads');
				}

				pushUploads();

				filePair.dest = (filePair.dest === '/') ? '' : filePair.dest;

				objects.push(_.defaults(filePair, filePairOptions));
			} else {

				if (filePair.params && !isValidParams(filePair.params)) {
					grunt.warn('"params" can only be ' + put_params.join(', '));
				} else {
					filePair.params = _.defaults(filePair.params || {}, options.params);
					_.defaults(filePair, filePairOptions);

					filePair.src.forEach(function (src) {

						// Prevents creating empty folders
						if (!grunt.file.isDir(src)) {

							if (_.last(filePair.dest) === '/') {
								dest = (is_expanded) ? filePair.dest : unixifyPath(path.join(filePair.dest, src));
							} else {
								dest = filePair.dest;
							}

							// '.' means that no dest path has been given (root). Nothing to create there.
							if (dest !== '.') {

								uploads.push(_.defaults({
									need_upload: true,
									src: src,
									dest: dest
								}, filePair));
							}
						}
					});
				}
			}
		});

		pushUploads();

		// Will list *all* the content of the bucket given in options
		// Recursively requests the bucket with a marker if there's more than
		// 1000 objects. Ensures uniqueness of keys in the returned list. 
		var listObjects = function (prefix, callback, marker, contents) {

			var search = {
				Prefix: prefix,
				Bucket: options.bucket
			};

			if (marker) {
				search.Marker = marker;
			}

			s3.listObjects(search, function (err, list) {

				if (!err) {

					var objects = (contents) ? contents.concat(list.Contents) : list.Contents;

					if (list.IsTruncated) {
						var new_marker = _.last(list.Contents).Key;
						listObjects(prefix, callback, new_marker, objects);
					} else {
						callback(_.uniq(objects, function (o) {
							return o.Key;
						}));
					}
				} else {
					grunt.fatal('Failed to list content of bucket ' + options.bucket + '\n' + err);
				}
			});
		};

		var deleteObjects = function (task, callback) {

			grunt.log.writeln('Deleting the content of ' + getObjectURL(task.dest).cyan);

			// List all the objects using dest as the prefix
			listObjects(task.dest, function (to_delete) {

				// List local content if it's a differential task
				var local_files = (task.differential) ? grunt.file.expand({
					cwd: task.cwd
				}, ['**']) : [];

				_.each(to_delete, function (o) {

					o.need_delete = true;
					o.excluded = task.exclude && grunt.file.isMatch(task.exclude, o.Key);

					if (task.exclude && task.flipExclude) {
						o.excluded = !o.excluded;
					}

					if (task.differential && !o.excluded) {
						// Exists locally or not (remove dest in the key to get the local path)
						o.need_delete = local_files.indexOf(getRelativeKeyPath(o.Key, task.dest)) === -1;
					}
				});

				// Just list what needs to be deleted so it can be sliced if necessary
				var delete_list = _.filter(to_delete, function (o) {
					return o.need_delete && !o.excluded;
				});

				if (options.debug) {
					callback(null, to_delete);
				} else if (delete_list.length > 0) {

					// deleteObjects requests can only take up to 1000 keys
					// If we are deleting more than a 1000 objects, we need slices
					var slices = Math.ceil(delete_list.length / 1000);
					var errors = [];
					var failed = [];
					var deleted = [];
					var calls = 0;

					var end = function (err, data) {

						if (err) {
							errors.push(err);
							data = data || {};
							failed = failed.concat(data.Errors || []);
						} else {
							deleted = deleted.concat(data.Deleted);
							grunt.log.write('.'.green);
						}

						if (++calls === slices) {
							if (errors.length > 0) {
								callback(JSON.stringify(errors), failed);
							} else {
								callback(null, to_delete);
							}
						}
					};

					var deleteSlice = function (i) {

						var start = 1000 * i;
						var slice = {
							Objects: _.map(delete_list.slice(start, start + 1000), function (o) {
								return {
									Key: o.Key
								};
							})
						};

						s3.deleteObjects({
							Delete: slice,
							Bucket: options.bucket
						}, function (err, data) {
							end(err, data);
						});
					};

					for (var i = 0; i < slices; i++) {
						deleteSlice(i);
					}
				} else {
					callback(null, (to_delete.length > 0) ? to_delete : null);
				}
			});
		};

		var doDownload = function (object, callback) {

			if (options.debug || !object.need_download || object.excluded) {
				callback(null, false);
			} else if (object.stream) {
				grunt.file.mkdir(path.dirname(object.dest));

				var stream = fs.createWriteStream(object.dest);
				var s3_object = s3.getObject({
					Key: object.Key,
					Bucket: options.bucket
				}).createReadStream();

				stream.on('finish', function () {
					callback(null, true);
				});

				s3_object.on('error', function (err) {
					callback(err);
				});

				stream.on('error', function (err) {
					callback(err);
				});

				s3_object.pipe(stream);
			} else {
				s3.getObject({
					Key: object.Key,
					Bucket: options.bucket
				}, function (err, data) {
					if (err) {
						callback(err);
					} else {
						grunt.file.write(object.dest, data.Body);
						callback(null, true);
					}
				});
			}
		};

		var downloadObjects = function (task, callback) {

			grunt.log.writeln('Downloading the content of ' + getObjectURL(task.dest).cyan + ' to ' + task.cwd.cyan);

			// List all the objects using dest as the prefix
			listObjects(task.dest, function (to_download) {

				// List local content if it's a differential task
				var local_files = (task.differential) ? grunt.file.expand({
					cwd: task.cwd
				}, ['**']) : [];

				if (to_download.length === 0) {
					callback(null, null);
				} else {

					var download_queue = async.queue(function (object, downloadCallback) {

						var key = getRelativeKeyPath(object.Key, task.dest); // Remove the dest in the key to not duplicate the path with cwd
						object.dest = task.cwd + key;
						object.stream = task.stream;
						object.need_download = _.last(object.dest) !== '/'; // no need to write directories
						object.excluded = task.exclude && grunt.file.isMatch(task.exclude, object.Key);
						object.gzip = task.gzip && !grunt.file.isMatch({
							matchBase: true
						}, task.excludedFromGzip, object.Key);

						if (task.exclude && task.flipExclude) {
							object.excluded = !object.excluded;
						}

						if (task.differential && object.need_download && !object.excluded) {
							var local_index = local_files.indexOf(key);

							// If file exists locally we need to check if it's different
							if (local_index !== -1) {

								// Check md5 and if file is older than server file
								var check_options = {
									file_path: object.dest,
									server_hash: object.ETag,
									server_date: object.LastModified,
									date_compare: 'older',
									gzip: object.gzip
								};

								isFileDifferent(check_options, function (err, different) {
									if (err) {
										downloadCallback(err);
									} else {
										object.need_download = different;
										setImmediate(doDownload, object, downloadCallback);
									}
								});
							} else {
								setImmediate(doDownload, object, downloadCallback);
							}
						} else {
							setImmediate(doDownload, object, downloadCallback);
						}

					}, options.downloadConcurrency);

					download_queue.drain = function () {

						callback(null, to_download);
					};

					download_queue.push(to_download, function (err, downloaded) {

						if (err) {
							grunt.fatal('Failed to download ' + getObjectURL(this.data.Key) + '\n' + err);
						} else {
							var dot = (downloaded) ? '.'.green : '.'.yellow;
							grunt.log.write(dot);
						}
					});
				}
			});
		};

		var doUpload = function (object, callback) {

			if (object.need_upload && !options.debug) {

				var type = options.mime[object.src] || object.params.ContentType || mime.lookup(object.src);
				var upload = _.defaults({
					ContentType: type,
					Key: object.dest,
					Bucket: options.bucket,
					ACL: options.access
				}, object.params);

				if (object.gzip) {
					upload.ContentEncoding = 'gzip';
				}

				var wrapped_callback = function () {
					s3.putObject(upload, function (err) {
						callback(err, true);
					});
				};

				console.log('object: ' + JSON.stringify(object, null, 4));
				if (object.stream) {
					var file_stream = fs.createReadStream(object.src);
					if (object.gzip) {
						// Can't use putObject with gzip stream - need to know length in advance

						var chunks = [];
						file_stream = file_stream.pipe(zlib.createGzip());
						file_stream.on('data', function (chunk) {
							chunks.push(chunk);
						});
						file_stream.on('end', function () {
							upload.Body = Buffer.concat(chunks);
							wrapped_callback();
						});
					} else {
						upload.Body = file_stream;
						wrapped_callback();
					}
				} else {
					var file_contents = grunt.file.read(object.src, {
						encoding: null
					});
					if (object.gzip) {
						zlib.gzip(file_contents, function (err, compressed) {
							if (err) {
								callback(err);
							} else {
								upload.Body = compressed;
								wrapped_callback();
							}
						});
					} else {
						upload.Body = file_contents;
						wrapped_callback();
					}
				}

			} else {
				callback(null, false);
			}
		};

		var uploadObjects = function (task, callback) {

			grunt.log.writeln('Uploading to ' + getObjectURL(task.dest).cyan);

			var startUploads = function (server_files) {

				var upload_queue = async.queue(function (object, uploadCallback) {

					object.gzip = object.gzip && !grunt.file.isMatch({
						matchBase: true
					}, object.excludedFromGzip, object.src);

					var server_file;
					if (object.differential && (server_file = _.find(server_files, {
							Key: object.dest
						}))) {
						isFileDifferent({
								file_path: object.src,
								server_hash: server_file.ETag,
								gzip: object.gzip
							},
							function (err, different) {
								object.need_upload = different;
								setImmediate(doUpload, object, uploadCallback);
							}
						);
					} else {
						setImmediate(doUpload, object, uploadCallback);
					}

				}, options.uploadConcurrency);

				upload_queue.drain = function () {

					callback(null, task.files);
				};

				upload_queue.push(task.files, function (err, uploaded) {

					if (err) {
						grunt.fatal('Failed to upload ' + this.data.src + ' with bucket ' + options.bucket + '\n' + err);
					} else {
						var dot = (uploaded) ? '.'.green : '.'.yellow;
						grunt.log.write(dot);
					}
				});
			};

			var unique_dests = _(task.files)
				.filter('differential')
				.pluck('dest')
				.compact()
				.map(path.dirname)
				.sort()
				.uniq(true)
				.reduce(function (res, dest) {
					var last_path = res[res.length - 1];
					if (!last_path || dest.indexOf(last_path) !== 0) {
						res.push(dest);
					}
					return res;
				}, []);

			if (unique_dests.length) {
				async.mapLimit(unique_dests, options.uploadConcurrency, function (dest, callback) {
					listObjects(dest, function (objects) {
						callback(null, objects);
					});
				}, function (err, objects) {
					if (err) {
						callback(err);
					} else {
						var server_files = Array.prototype.concat.apply([], objects);
						startUploads(server_files);
					}
				});
			} else {
				startUploads([]);
			}
		};

		var queue = async.queue(function (task, callback) {

			if (task.action === 'delete') {
				deleteObjects(task, callback);
			} else if (task.action === 'download') {
				downloadObjects(task, callback);
			} else {
				uploadObjects(task, callback);
			}
		}, 1);

		queue.drain = function () {

			_.each(objects, function (o) {

				if (o.action === 'delete') {
					grunt.log.writeln(o.deleted.toString().green + '/' + o.nb_objects.toString().green + ' objects deleted from ' + (options.bucket + '/' + o.dest).green);
				} else if (o.action === 'download') {
					grunt.log.writeln(o.downloaded.toString().green + '/' + o.nb_objects.toString().green + ' objects downloaded from ' + (options.bucket + '/' + o.dest).green + ' to ' + o.cwd.green);
				} else {
					grunt.log.writeln(o.uploaded.toString().green + '/' + o.nb_objects.toString().green + ' objects uploaded to bucket ' + (options.bucket + '/').green);
				}
			});

			if (options.debug) {
				grunt.log.writeln('\nThe debug option was enabled, no changes have actually been made'.yellow);
			}

			done();
		};

		queue.push(objects, function (err, res) {
			var object_url = getObjectURL(this.data.dest);

			if (this.data.action === 'delete') {
				if (err) {
					if (res && res.length > 0) {
						grunt.log.writeln('Errors (' + res.length.toString().red + ' objects): ' + _.pluck(res, 'Key').join(', ').red);
					}

					grunt.fatal('Deletion failed\n' + err.toString());
				} else {
					if (res && res.length > 0) {
						grunt.log.writeln('\nList: (' + res.length.toString().cyan + ' objects):');

						var deleted = 0;

						_.each(res, function (file) {

							if (file.need_delete && !file.excluded) {
								deleted++;
								grunt.log.writeln('- ' + file.Key.cyan);
							} else {
								var sign = (file.excluded) ? '! ' : '- ';
								grunt.log.writeln(sign + file.Key.yellow);
							}
						});

						this.data.nb_objects = res.length;
						this.data.deleted = deleted;
					} else {
						grunt.log.writeln('Nothing to delete');
						this.data.nb_objects = 0;
						this.data.deleted = 0;
					}
				}
			} else if (this.data.action === 'download') {
				if (err) {
					grunt.fatal('Download failed\n' + err.toString());
				} else {
					if (res && res.length > 0) {
						grunt.log.writeln('\nList: (' + res.length.toString().cyan + ' objects):');

						var task = this.data;
						var downloaded = 0;

						var printFile = function (file, color, sign) {
							grunt.log.writeln('- ' + getObjectURL(file.Key)[color] + ' ' + sign + ' ' + (task.cwd + getRelativeKeyPath(file.Key, task.dest))[color]);
						};

						_.each(res, function (file) {

							if (file.need_download && !file.excluded) {
								downloaded++;
								printFile(file, 'cyan', '->');
							} else {
								var sign = (file.excluded) ? '=/=' : '===';
								printFile(file, 'yellow', sign);
							}
						});

						this.data.nb_objects = res.length;
						this.data.downloaded = downloaded || 0;
					} else {
						grunt.log.writeln('Nothing to download');
						this.data.nb_objects = 0;
						this.data.downloaded = 0;
					}
				}
			} else {
				if (err) {
					grunt.fatal('Upload failed\n' + err.toString());
				} else {
					grunt.log.writeln('\nList: (' + res.length.toString().cyan + ' objects):');

					var uploaded = 0;

					_.each(res, function (file) {

						if (file.need_upload) {
							uploaded++;
							grunt.log.writeln('- ' + file.src.cyan + ' -> ' + (object_url + file.dest).cyan);
						} else if (!options.displayChangesOnly) {
							grunt.log.writeln('- ' + file.src.yellow + ' === ' + (object_url + file.dest).yellow);
						}
					});

					this.data.nb_objects = res.length;
					this.data.uploaded = uploaded;
				}
			}

			grunt.log.writeln();
		});
	});

	var unixifyPath = function (filepath) {

		if (process.platform === 'win32') {
			return filepath.replace(/\\/g, '/');
		} else {
			return filepath;
		}
	};
};